lib/internal/http2/core.js

```
@@ -1519,36 +1704,43 @@ function processHeaders(headers) {
   return headers;
 }
 
-function processRespondWithFD(fd, headers, offset = 0, length = -1,
+function processRespondWithFD(self, fd, headers, offset = 0, length = -1,
                               streamOptions = 0) {
-  const state = this[kState];
-  state.headersSent = true;
+  const state = self[kState];
+  state.flags |= STREAM_FLAGS_HEADERS_SENT;
+
+  const headersList = mapToHeaders(headers, assertValidPseudoHeaderResponse);
+  if (!Array.isArray(headersList)) {
+    self.destroy(headersList);
+    return;
+  }
+
 
   // Close the writable side of the stream
-  this.end();
+  self.end();
 
-  const ret = this[kHandle].respondFD(fd, headers,
+  const ret = self[kHandle].respondFD(fd, headersList,
                                       offset, length,
                                       streamOptions);
 
   if (ret < 0) {
-    const err = new NghttpError(ret);
-    process.nextTick(emit, this, 'error', err);
+    self.destroy(new NghttpError(ret));
     return;
   }
   // exact length of the file doesn't matter here, since the
   // stream is closing anyway — just use 1 to signify that
   // a write does exist
-  trackWriteState(this, 1);
+  trackWriteState(self, 1);
 }
 
 function doSendFD(session, options, fd, headers, streamOptions, err, stat) {
-  if (this.destroyed || session.destroyed) {
-    abort(this);
+  if (err) {
+    this.destroy(err);
     return;
   }
-  if (err) {
-    process.nextTick(emit, this, 'error', err);
+  if (this.destroyed || this.closed) {
+    // TODO(jasnell): emit error here instead?
+    abort(this);
```

silently ignoring things is not nice, I'd do the error here.